cram/cram_index.c

```diff
@@ -86,6 +86,37 @@ static void dump_index(cram_fd *fd) {
 }
 #endif
 
+// Thread a linked list through the nested containment list.
+// This makes navigating it and finding the "next" index entry
+// trivial.
+static cram_index *link_index_(cram_index *e, cram_index *e_last) {
+    int i;
+    if (e_last)
+        e_last->e_next = e;
+
+    // We don't want to link in the top-level cram_index with
+    // offset=0 and start/end = INT_MIN/INT_MAX.
+    if (e->offset)
+        e_last = e;
```

Comment:
It is possible to avoid the recursion, by spotting the ladder and walking along it...
```suggestion
    if (e->offset) {
        e_last = e;

        // Prevent excess recursion on very deep single-child nestings.
        // This can especially happen on unmapped, unplaced reads, which all
        // pile up at position 0.
        while (e->nslice == 1) {
            e = &e->e[0];
            e_last->e_next = e;
            e_last = e;
        }
```

Reply:
I'm trying to recall why the nested containment list was used in the first place.  I think treating the index as an array that we can binary search on only really works if we ensure the list of start positions are always sorted, and similarly for the end positions.  This is ensured by using NCList here, or the recursive binning in BAM (which could be viewed as a nested list too).  I think we should permit single-child nesting, but instead avoid making containments where start and end are equal (as that's still sorted by both ends).

Unlike BAM where we produce a hit list of multiple bins to scan, with seeking between them.

I think the intention here with containment lists was mixing very long reads with very short reads would return the first large outer container, and then subsequent offsets within a containment list would be yielded by calling the query again using the "from" parameter.  (That's rewritten and fixed here, but not in the way originally intended of recursing down and doing a second level of binary searching).  Instead we just found the first container and then did reading and/or seeking to consume or skip containers until we're past the end of the range.  It's almost always the same amount of I/O, but in the extreme case it's possible we could be seeking past numerous containers.  In practice though it's sufficient and fast enough to not worry about it.  CRAM can get away with it due to the container structure permitting seekage, unlike BAM.

So actually it may not matter if both start/end are sorted (start alone is sufficient if we then skip from that point on), but I'd still like to keep that as a possibility.

Reply:
Ah the lack of context in the quoted code above confused me.  You're not changing the index loading to avoid deeply nesting the list, but simply avoiding recursion in the link_index function.  Yeah that's safer, but I think unnecessary too.

We should just change the index loading code to work better and not create containments for exact matching start/end on unmapped data.